connectors/livelink/connector/src/main/java/org/apache/manifoldcf/livelink/LLSERVER.java

```diff
@@ -0,0 +1,345 @@
+*/
+public class LLSERVER
+{
+  public static final String _rcsid = "@(#)$Id: LLSERVER.java 988245 2010-08-23 18:39:35Z kwright $";
+
+  private final boolean useHttp;
+  private final boolean useSSL;
+  private final String LLServer;
+  private final int LLPort;
+  private final String LLUser;
+  private final String LLPwd;
+  private final String httpCgiPath;
+  private final String httpNtlmDomain;
+  private final String httpNtlmUser;
+  private final String httpNtlmPassword;
+  private final IKeystoreManager keystore;
+  
+  private LLSession session = null;
+  private File certFolder = null;
+
+
+  public LLSERVER(boolean useHttp, boolean useSSL, String server, int port, String user, String pwd,
+    String httpCgiPath, String httpNtlmDomain, String httpNtlmUser, String httpNtlmPassword,
+    IKeystoreManager keystoreManager)
+    throws ManifoldCFException
+  {
+    this.useHttp = useHttp;
+    this.useSSL = useSSL;
+    LLServer = server;
+    LLPort = port;
+    LLUser = user;
+    LLPwd = pwd;
+    this.httpCgiPath = httpCgiPath;
+    this.httpNtlmDomain = httpNtlmDomain;
+    this.httpNtlmUser = httpNtlmUser;
+    this.httpNtlmPassword = httpNtlmPassword;
+    this.keystore = keystoreManager;
+
+    connect();
+  }
+
+  private void connect()
+    throws ManifoldCFException
+  {
+    try
+    {
+      LLValue configuration;
+
+      if (useHttp)
+      {
+        boolean useNTLM;
+        String userNameAndDomain;
+
+        if (httpNtlmDomain != null && httpNtlmUser != null)
+        {
+          useNTLM = true;
+          userNameAndDomain = httpNtlmDomain + "\\" + httpNtlmUser;
+        }
+        else
+        {
+          useNTLM = false;
+          userNameAndDomain = httpNtlmUser;
+        }
+        configuration = new LLValue();
+        configuration.setAssoc();
+        configuration.add("Encoding","UTF-8");
+        configuration.add("LivelinkCGI", httpCgiPath);
+        if (userNameAndDomain != null)
+        {
+          configuration.add("HTTPUserName", userNameAndDomain);
+          configuration.add("HTTPPassword", httpNtlmPassword);
+        }
+        if (useNTLM)
+          configuration.add("EnableNTLM", LLValue.LL_TRUE);
+        else
+          configuration.add("EnableNTLM", LLValue.LL_FALSE);
+
+        if (useSSL)
+        {
+          configuration.add("HTTPS", LLValue.LL_TRUE);
+          // Create the place to put the certs
+          createCertFolder();
+          if (keystore != null)
+          {
+            // Now, write the certs themselves
+            String[] aliases = keystore.getContents();
+            for (String alias : aliases)
+            {
+              java.security.cert.Certificate cert = keystore.getCertificate(alias);
+              byte[] certData = cert.getEncoded();
+              File fileName = new File(certFolder,ManifoldCF.safeFileName(alias) + ".cer");
+              OutputStream fos = new FileOutputStream(fileName);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *fos*. The resource is referenced in statements at the following lines: 138, 142, 148, 153, 157, 161, 166. The resource closure statement is at line: 167. There are other execution paths that do not contain closure statements, e.g., when exception is thrown by OutputStream.flush. Either a) close *fos* in a try-finally block or b) close the resource by declaring *fos* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).